### PR TITLE
[feature] (Nereids) add rule to merge consecutive project nodes

### PR DESCRIPTION
**Pull request by englefly**
# Proposed changes
Merge consecutive project nodes. For example:
 logical plan tree:
 ```
                project(a)
                   |
                 project(a,b)
                   |
                 project(a, b, c)
                   |
                 scan
```
  transformed to:
```
                 project(a)
                    |
                 scan
```
Issue Number: close #xxx

## Checklist(Required)

1. Type of your changes:
    - [ ] Improvement
    - [ ] Fix
    - [ ] Feature-WIP
    - [ ] Feature
    - [ ] Doc
    - [ ] Refator
    - [ ] Others: 
2. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
3. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
6. Are there any changes that cannot be rolled back:
    - [ ] Yes
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by morrySnow — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ *                   |
+ *                 scan
+ */
+
+public class MergeConsecutiveProjects extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalProject(logicalProject()).then(project -> {
+            List<NamedExpression> projectExpressions = project.getProjects();
+            LogicalProject childProject = project.child();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java:60]:
```suggestion
            LogicalProject<GroupPlan> childProject = project.child();
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                    new ExpressionReplacer<Map<ExprId, Alias>>();
+            projectExpressions = projectExpressions.stream()
+                    .map(e -> replacer.visit(e, childAliasMap))
+                    .map(NamedExpression.class::cast)
+                    .collect(Collectors.toList());
+            return new LogicalProject(projectExpressions, (Plan) childProject.children().get(0));
+        }).toRule(RuleType.MERGE_CONSECUTIVE_PROJECTS);
+    }
+
+    private class ExpressionReplacer<M>
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java:77]: move out ExpressionReplacer from AggregateDisassemble to tree/expressions/visitor and use it here

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    public Rule build() {
+        return logicalProject(logicalProject()).then(project -> {
+            List<NamedExpression> projectExpressions = project.getProjects();
+            LogicalProject childProject = project.child();
+            List<NamedExpression> childProjectExpressions = childProject.getProjects();
+            HashMap<ExprId, Alias> childAliasMap = new HashMap<ExprId, Alias>();
+            AliasExtractor extractor = new AliasExtractor();
+            for (NamedExpression expression : childProjectExpressions) {
+                extractor.visit(expression, childAliasMap);
+            }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java:66]:
```suggestion
            Map<Expression, Expression> sMap = childProjectExpressions.stream().collect(Collectors.toMap(
                    NamedExpression::toSlot, e -> e instanceof SlotReference ? e : e.child(0)));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/MergeConsecutiveProjects.java:77]: They have different logical.
In this case, the key is ExprId, while in AggregateDisassemble the key is expression.
Comparison by exprId is more efficient than comparison by expression tree.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ExpressionReplacer.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.apache.doris.nereids.trees.expressions.Expression;
+
+import java.util.Map;
+
+/**
+ * replace expr nodes by substitutionMap
+ */
+public class ExpressionReplacer
+        extends DefaultExpressionRewriter<Map<Expression, Expression>> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ExpressionReplacer.java:28]: add singleton back
```
private static final ExpressionReplacer INSTANCE = new ExpressionReplacer();
```

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ExpressionReplacer.java:28]: ok

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ExpressionReplacer.java:28]: done
but it is public now

**Review by morrySnow — DISMISSED** [review]

**Review by 924060929 — DISMISSED** [review]
LGTM.

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**